### PR TITLE
PR for Issue #471 Added Thread Classes: gsl::raii_thread gsl::detached_thread

### PR DESCRIPTION
**Pull request by galik**
This PR adds <gsl/gsl_thread> and tests/thread_tests.cpp
Fixes for Issue #471 

**Notes:**
I have copied the interface of `std::thread` except for `std::thread::detach`. The reason being that `gsl::detached_thread` is already detached and `gsl::raii_thread` should not be detached by design.

### CODE REVIEW

File: include/gsl/gsl_thread [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+        t.detach();
+    }
+
+    bool joinable() const { return t.joinable(); }
+
+    std::thread::id get_id() const noexcept { return t.get_id(); }
+
+    std::thread::native_handle_type native_handle() { return t.native_handle(); }
+
+    void join() { t.join(); }
```
Comment (mlalic) [inline include/gsl/gsl_thread]: If the `std::thread` is actually detached, this will throw. I don't think there's a point in exposing a method that will _always_ just throw no matter what.

File: include/gsl/gsl_thread [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+    raii_thread(raii_thread const&) = delete;
+    raii_thread(raii_thread&& other) : t(std::move(other.t)) {}
+
+    raii_thread(std::thread const&) = delete;
+    raii_thread(std::thread&& other) noexcept : t(std::move(other)) {}
+
+    raii_thread& operator=(raii_thread const&) = delete;
+    raii_thread& operator=(raii_thread&& other) noexcept
+    {
+        t = std::move(other.t);
```
Comment (mlalic) [inline include/gsl/gsl_thread]: This `std::terminate`s if `this->t` hasn't been joined. This isn't in the spirit of an RAII wrapper. It should likely first properly "reset" `this` and then move the `rhs` into it, similarly to what other RAII managers, like `unique_ptr` would do.

A check `if (this != &rhs)` on top of it would be needed too.

Comment (galik) [inline include/gsl/gsl_thread]: When you say "reset" `this` do you mean perform a `join()`? Because I am not entirely sure about that. 

I could be wrong but my feeling is that the main use for this function is when you are *moving* into a default initialized `raii_thread`. For example when returning a `raii_thread` from a function or accepting one as a parameter.

If the client actually assigns a `raii_thread` to an already running one then performing a join might introduce hidden, unexpected serialization delays and I suspect it would rarely be the intention to *move* a running `raii_thread`onto another already running one.

Would that not always be an error?

I am willing to be persuaded either way. I think I would generally rather it trigger a terminal error than silently join because (with an `raii_thread`) I would probably always want to do the join explicitly using the natural scope.

**Comment by hsutter** [comment]
The main thing is to get `raii_thread` whose destructor unconditionally joins, which is what `std::thread` should have done. That part looks fine. Then we can recommend using `raii_thread` everywhere.

I'm less positive about adding a `detached_thread`... I'm not sure we should be encouraging detached threads.

Suggestion: Could we separate the PR into two so that we can adopt the `raii_thread` and further discuss the advisability of `detached_thread`?

**Comment by galik** [comment]
See new PR #504
